docs/learning/guidelines/release-tags-guidelines.md

```diff
@@ -1,66 +1,205 @@
- * @extensions
- */
+/** @public */
+export class Thing {
+  /** @internal */
+  public close(): void {
+    // implementation goes here
+  }
+}
+
 ```
+
+Add to `src/internal/Symbols.ts` the following, if a `_close` symbol doesn't already exist (keeping the list of symbols sorted alphabetically):
+
+```ts
+export const _close = Symbol.for("close_INTERNAL_ONLY_DO_NOT_USE");
+```
+
+Update `Thing` to rename `close` to use the symbol, and deprecate the existing `close` method:
+
+```ts
+/** @public */
+export class Thing {
+  /** @internal */
+  public [_close]: () => void {
+    // implementation goes here
+  }
+  
+  /** @internal
+   * @deprecated in 4.8. The thing will automatically be closed when it is no longer in use. This API is for internal use only and will soon be removed.
+   */
+  public close(): void {
+    this[_close]();
+  }
+}
+```
+
+Update all callers of `close` in itwinjs-core to use `[_close]` instead.
+
+The symbols in `Symbols.ts` are not exported from `core-backend.ts`, so they are inaccessible to packages outside of itwinjs-core. We preserve the original `@internal` API to avoid introducing a breaking API change for external callers (who should not be using it in the first place). The deprecation tag will warn them to fix their code to remove their dependency on the internal API. If we were reasonably certain no one was depending on the `close` method, we could simply delete it.
+
+### Transition plan
+
+In iTwin.js 5.0, we will be able to begin enforcing the internal API policy without regard for breaking changes.
+
+Currently, iTwin.js publishes both CommonJS modules (`/lib/cjs/`) and ESModules (`/lib/esm/`), which prevents one package from importing a top-level API from another package using a relative path because it can't know which type of module it should import from. Also, some top-level `@internal` APIs are known or suspected to be used by code outside of itwinjs-core.
+
+- In 4.x (using the `core-backend` example above), we will export the contents of `cross-package.ts` from `core-backend.ts`.
+- In 5.0, we will standardize on ESModules, allowing, e.g., core-backend to import `InternalApi` from core-common using `import { InternalApi } from "@itwin/core-common/lib/InternalApi";`. We can then delete `cross-package.ts`.
+
+Currently, it's possible for an app to (mis)configure their dependencies such that they end up with multiple versions of single core package - and hence, multiple independent copies of each `Symbol` defined in `Symbols.ts`.
+- In 4.x, we use `Symbol.for` to define those symbols so that they are looked up by name in a global registry, preventing duplication.
+- In 5.0, we will prohibit taking a dependency on multiple versions of the same core package (which can lead to all sorts of other problems), and switch to using the `Symbol` constructor, which does not register the symbol in the global registry. This will make it impossible for anyone outside the package to look up the symbol by its name.
```

Comment:
Are we planning to do a runtime validation or what's the proposal here? maybe I missed that in the GH issue.

Reply:
I don't believe anyone gave specifics during our meeting about how they would enforce this, only that it should be enforced. I think there's a run-time enforcement for core-frontend right now that checks for the presence of a property on `window`. @aruniverse @wgoehrig?

Reply:
Correct, we would look to copy the approach from [frontend](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/extension/ExtensionRuntime.ts#L15-L17) into other packages

Reply:
@aruniverse sounds good. Let's add that as its own GH issue for 5.0 release.